### ISSUE TITLE
Differentiating between Fixedwing, Multicopter and VTOL (and others?)

### ISSUE THREAD

**Issue by JonasVautherin** [mavlink/MAVSDK-Proto#168]
Right now, MAVSDK is developed and tested against multicopters. Some functions are generic and will work with any MAVLink system (e.g. receiving telemetry messages). But some are not: `action.land()` makes no sense for a rover, and requires more information for a fixedwing (e.g. direction of the landing).

There had been discussions about having different plugins for different types. E.g. `FwMission`, `McMission` and `VtolMission`, possibly all extending a `Mission` base class. But I believe this will look weird from the user's perspective. Let's assume that a multicopter has `action.land()`, but a fixedwing needs `action.land(param1, param2)`. The code will need some casting in static-typed languages. In pseudocode:

```
Action().land() // This won't work, because there is no common 'land()' in the base class

((McAction)Action()).land() // landing for a multicopter
((FwAction)Action()).land(param1, param2) // landing for a fixedwing
```

On top of that, if the user makes a wrong cast, they will have to deal with the consequences (e.g. catch an exception?) and therefore, probably they will want to do more something like this (at least I would):

```
if (Drone.isMulticopter()) {
    ((McAction)Action()).land() // landing for a multicopter
} else if (Drone.isFixedwing()) {
    ((FwAction)Action()).land(param1, param2) // landing for a fixedwing
}
```

But at this point, why would we need a cast? If we offer `Drone.isMulticopter()`, maybe we could just have everything in the same class.

My suggestion here would be the following:
* We offer some kind of `Drone.isMulticopter()`/`Drone.isRover()` interface to the user.
* Plugins may have functions for different platforms, i.e. `mission.proto` is here for all platforms, and we do not create `mission_fixedwing.proto` and similar.
* If the user calls the wrong function (e.g. `drone.action.transition_to_multicopter()` on a fixedwing), then the function's implementation deals with that (e.g. by ignoring the call, showing a warning, or throwing an exception).

As a result, still with the `land()` pseudo-code, it would look like this:

```
// This works
if (Drone.isMulticopter()) {
    drone.action.land()
} else if (Drone.isFixedwing()) {
    drone.action.land(param1, param2)
}

// This fails (throws an exception, or something), because a fixedwing can't land with the land() function (it needs more info)
if (Drone.isFixedwing()) {
    drone.action.land()
}

// This may or may not fail, depending on our implementation choice, because it seems safe to ignore the fixedwing parameters when landing a multicopter
if (Drone.isMulticopter()) {
    drone.action.land(param1, param2)
}
```

@hamishwillee @julianoes: FYI

**Comment by JonasVautherin** [comment]
Note: this is done similarly on some cross-platform frameworks like Flutter:

```
if (Platform.isAndroid) {
  // Android-specific code
} else if (Platform.isIOS) {
  // iOS-specific code
}
```

**Comment by iwishiwasaneagle** [comment]
Hi @JonasVautherin, I do like this. Prevents a ton of copy/pasted code which in return will improve code maintenance. 

My main gripe with it is how `mission` currently is seemingly only for multi-rotors. Take `GetReturnToLaunchAfterMission` for example, this applies to mutli-rotors, and VTOLs but not really for FW in the same way (holding pattern above launch, rather than a land). I only mention those 3 because I have no experience with the other vehicles.

Also, at the point where you've written `land`, `take-off`, etc. for each vehicle is there any difference between the current implementation and what was suggested?

**Comment by JonasVautherin** [comment]
> My main gripe with it is how mission currently is seemingly only for multi-rotors.

Yes, we actually really appreciate your insights for fixedwing. I personally don't have any experience with them. With your input, we should be able to make it nicer for fixedwing (and vtol, to some extent).

> Also, at the point where you've written land, take-off, etc. for each vehicle is there any difference between the current implementation and what was suggested?

The main difference I see is that if you make multiple proto files, you end up having multiple classes, and users need a cast in their code to get access to the class-specific methods (not to mention that we would have a hard time writing the `drone` helper in that case). If you have all the methods defined in one proto file, then they all end up in one class without needing a cast. Of course, users still have to make sure they don't call the wrong method (e.g. `transition_to_fixedwing()` will fail when controlling a multicopter), but I find the code more readable this way. 

Actually, I do believe that the user code would get quite convoluted with the per-platform plugins and the whole casting situation :sweat_smile:.

**Comment by iwishiwasaneagle** [comment]
> The main difference I see is that if you make multiple proto files, you end up having multiple classes, and users need a cast in their code to get access to the class-specific methods (not to mention that we would have a hard time writing the `drone` helper in that case). If you have all the methods defined in one proto file, then they all end up in one class without needing a cast. Of course, users still have to make sure they don't call the wrong method (e.g. `transition_to_fixedwing()` will fail when controlling a multicopter), but I find the code more readable this way.

I agree with you, but surely if there's `mission_mc`, `mission_fw`, etc the user will import the plugin that they require removing the need for any casting? However, I think you're also taking about the `action` plugin right? For `action` your solution definitely feels intuitive - I've done loads of flutter programming and the `isX` is very useful. Maybe even a `getPlatform` returning an enum for use in a switch? Then `isX` simply becomes a wrapper for `getPlatform`

> Actually, I do believe that the user code would get quite convoluted with the per-platform plugins and the whole casting situation sweat_smile.
Yeah this is very true. If you do `mission_vtol`, `mission_fixed_wing`, etc. where do you stop? Currently there is a nice distinction between the various plugins and the per-platform plugins isn't very scalable.

**Comment by JonasVautherin** [comment]
> the user will import the plugin that they require removing the need for any casting?

I don't think we can count on that with our `Drone` wrapper (it's nice that users can do `drone.action.<something>`, but then they don't `import Action`, they just `import System`).

But more than that, if you wanted to write a script that works for both VTOL and multicopter, then you would anyway use the base class and make casts for platform-specific calls. So I don't think you can get away without ever casting.

**Comment by iwishiwasaneagle** [comment]
> > the user will import the plugin that they require removing the need for any casting?
> 
> I don't think we can count on that with our `Drone` wrapper (it's nice that users can do `drone.action.<something>`, but then they don't `import Action`, they just `import System`).
> 
> But more than that, if you wanted to write a script that works for both VTOL and multicopter, then you would anyway use the base class and make casts for platform-specific calls. So I don't think you can get away without ever casting.

Okay that makes sense to be fair.

So (talking about `mission` here) you are proposing that there are multiple `land`s, `takeoff`s etc? How does that translate into other languages that don't support function overloading, such as python?

**Comment by JonasVautherin** [comment]
> So (talking about mission here) you are proposing that there are multiple lands, takeoffs etc?

Yes. But only when it's needed. Let's take some examples:

* Say MAVLink had two messages to takeoff: `TAKEOFF_MC` and `TAKEOFF_VTOL`, both taking the same arguments. Then MAVSDK would just have `takeoff()`, and in the implementation we would call `TAKEOFF_MC` or `TAKEOFF_VTOL` (but the user would not realize it).
* Now it could be that a multicopter needs `land()`, and a fixedwing needs `land(param1, param2)`. In that case, we implement both, and make sure that `land()` crashes (in an elegant way) if called for a fixedwing.
* Finally, maybe all platforms are fine with exactly the same `arm()` implementation, and therefore we don't change anything.

> How does that translate into other languages that don't support function overloading, such as python?

Very good point! I don't know :laughing:. What about something like [this](https://stackoverflow.com/a/57726675/1368342)?

**Comment by iwishiwasaneagle** [comment]
> Yes. But only when it's needed. Let's take some examples:
>   -  Say MAVLink had two messages to takeoff: TAKEOFF_MC and TAKEOFF_VTOL, both taking the same arguments. Then MAVSDK would just have takeoff(), and in the implementation we would call TAKEOFF_MC or TAKEOFF_VTOL (but the user would not realize it).
>   -  Now it could be that a multicopter needs land(), and a fixedwing needs land(param1, param2). In that case, we implement both, and make sure that land() crashes (in an elegant way) if called for a fixedwing.
>   -  Finally, maybe all platforms are fine with exactly the same arm() implementation, and therefore we don't change anything.

Yeah that makes sense to me, and like you said, `arm()` for example then doesn't need to be re-written X times.

>Very good point! I don't know. What about something like this?

Sure, that works for python but isn't the whole point for the code generation on-top of C++ that you can generate the code for almost anything? Just brainstorming, but would a `land(TYPE,PARAMS)` work? where `TYPE` is an enum of landing types, and params is a object that's been extended for various types? so for example base class is `LandParams` with LatLng, etc. and then you have `FixedWingLandParams` that extends it etc

**Comment by JonasVautherin** [comment]
> you can generate the code for almost anything?

Yes, from templates. So in the python templates, we can generate any python code we want. But now I'm actually wondering if overloading would work from protobuf (because at some point we need to define the function in the proto file)...  I don't think it would :confused:.

Unless we create our own syntax, e.g.:

```
rpc Land_overload1(LandRequestOverload1) returns(LandResponseOverload1) {}
rpc Land_overload2(LandRequestOverload2) returns(LandResponseOverload2) {}
```

Or find a way to encode the different types (`MulticopterLandParams` and `FixedWinfLandParams`) inside `LandRequest` :thinking:.

I need to think about how we can define that in the proto file :thinking: :thinking:.

**Comment by iwishiwasaneagle** [comment]
> > you can generate the code for almost anything?
> 
> Yes, from templates. So in the python templates, we can generate any python code we want. But now I'm actually wondering if overloading would work from protobuf (because at some point we need to define the function in the proto file)... I don't think it would confused.
> 
> Unless we create our own syntax, e.g.:
> 
> ```
> rpc Land_overload1(LandRequestOverload1) returns(LandResponseOverload1) {}
> rpc Land_overload2(LandRequestOverload2) returns(LandResponseOverload2) {}
> ```
> 
> Or find a way to encode the different types (`MulticopterLandParams` and `FixedWinfLandParams`) inside `LandRequest` thinking.
> 
> I need to think about how we can define that in the proto file thinking thinking.

Well I guess the OOP paradigm of extending is nice, but not necessary for what I suggested. The only nice thing it provides is easy reading and a good overview. 

For protobuf, I won't be much help sorry.

```python
class MulticopterLandParams:
    def __init__()(self, lat, lng, alt, descent_rate):
        self.lat = lat
        self.lng = lng
        self.alt = alt
        self.descent_rate = descent_rate

class FixedWingLandParams:
    def __init__()(self, lat, lng, alt, glide_slope):
        self.lat = lat
        self.lng = lng
        self.alt = alt
        self.glide_slope = glide_slope

def land(params):
    if(type(params) == MulticopterLandParams):
        # land with MulticopterLandParams as a Multicopter
    elif(type(params) == FixedWingLandParams)
        # land with FixedWingLandParams as a Fixed Wing UAV
```

**Comment by JonasVautherin** [comment]
I think that your code proposition above makes sense. So now I just need to find a way to encode function overloading in the proto files.

I'm currently looking at the [OneOf](https://developers.google.com/protocol-buffers/docs/proto3#oneof) feature. I'm thinking that something like this may work:

```
message LandRequest {
  oneof land_params {
    MulticopterLandParams multicopter_land_params = 1;
    FixedWingLandParams fixedwing_land_params = 2;
  }
}
```

I'll give it a try to see if I can extract the information we need from that.

About `FixedWingLandParams`, do you think that would be enough to land a fixedwing, or was that just an example for the discussion, but won't work in practice (my current understanding is that a fixedwing needs a mission to land). Just trying to find a real example that we could actually test against SITL once we have our prototype :-).

**Comment by iwishiwasaneagle** [comment]
> I think that your code proposition above makes sense. So now I just need to find a way to encode function overloading in the proto files.
> 
> I'm currently looking at the [OneOf](https://developers.google.com/protocol-buffers/docs/proto3#oneof) feature. I'm thinking that something like this may work:
> 
> ```
> message LandRequest {
>   oneof land_params {
>     MulticopterLandParams multicopter_land_params = 1;
>     FixedWingLandParams fixedwing_land_params = 2;
>   }
> }
> ```
> 
> I'll give it a try to see if I can extract the information we need from that.
> 
> About `FixedWingLandParams`, do you think that would be enough to land a fixedwing, or was that just an example for the discussion, but won't work in practice (my current understanding is that a fixedwing needs a mission to land). Just trying to find a real example that we could actually test against SITL once we have our prototype :-).

That looks good, plus it gives a good oversight of what's currently accepted rather than going for a generic `Object` input (if that's even allowed in proto haha)

Yes, just a concept. I'm currently trying to figure out some more specifics for FW landing. From the looks of things, it's simple enough to do under the `mission.land`. We'd have to convert the `FixedWingLandParams` to about 3 individual mission steps at a minimum for a good landing (see below, currently being done using `MissionRaw`). 

https://gist.github.com/iwishiwasaneagle/0c16e5c98e02c9b4eab1c7d83ff26a8d
```python
import asyncio
from mavsdk import System, Mission_raw

async def run():
    drone = System()
    await drone.connect(system_address="udp://:14540")

    print("Waiting for drone to connect...")
    async for state in drone.core.connection_state():
        if state.is_connected:
            print(f"Drone discovered with UUID: {state.uuid}")
            break
    # Landing with a 6deg slope calcualted by QGC
    items = []
    items.append(
        mission_raw.MissionItem( # MAV_CMD_DO_LAND_START 189
            0,2,189,1,1,0,0,0,0,0,0,0,0)
    )
    items.append( # MAV_CMD_NAV_LOITER_TO_ALT 31
        mission_raw.MissionItem(
            1,6,31,0,1,1,75,0,1,558683815,-42929235,40.0,0)
    )
    items.append( # MAV_CMD_NAV_LAND 21
        mission_raw.MissionItem(
            2,6,21,0,1,0,0,0,0,558688774,-42867694,0,0
        )
    )
    await drone.mission_raw.upload_mission(items)
    await drone.mission_raw.start_mission()

if __name__ == "__main__":
    loop = asyncio.get_event_loop()
    loop.run_until_complete(run())
```

**Comment by JonasVautherin** [comment]
Well, let's close that for now.
